### PR TITLE
Tear down thindevs as part of StratPool::teardown

### LINKED ISSUE #471: Need to properly sync filesysystems in StratPool::teardown()
Whatever is necessary to make sure that the user data is where it needs to be before thin pools and metadata vols go away must be done here.

### PR DESCRIPTION
**Pull request by agrover**
This should let tearing down the thinpool itself succeed.

fixes #471

Signed-off-by: Andy Grover <agrover@redhat.com>

### CODE REVIEW

**Comment by mulkieran** [comment]
What's against putting a ```teardown()``` method in ```StratFilesystem```?

**Comment by agrover** [comment]
sure let's try that, one sec

**Comment by agrover** [comment]
updated.

**Comment by mulkieran** [comment]
Why will tearing down a thindev fail if the filesystem that is on it is still mounted?

File: src/engine/strat_engine/pool.rs [hunk 1]
```diff
@@ -377,12 +377,13 @@ impl StratPool {
...
     /// in sync.
     pub fn teardown(self) -> EngineResult<()> {
-        // TODO: any necessary clean up of filesystems
-        if !self.filesystems.is_empty() {
-            return Err(EngineError::Engine(ErrorEnum::Busy,
-                                           format!("May be unsynced files on device.")));
-        }
         let dm = try!(DM::new());
+
+        // Note: Will fail if filesystems are still mounted
```
Comment (mulkieran) [inline src/engine/strat_engine/pool.rs]: Is this note really correct? Or will the teardown succeed but make some sort of mess?

File: src/engine/strat_engine/filesystem.rs [hunk 2]
```diff
@@ -86,6 +86,11 @@ impl StratFilesystem {
     pub fn thin_id(&self) -> ThinDevId {
         self.thin_dev.id()
     }
+
+    /// Tear down the filesystem.
```
Comment (mulkieran) [inline src/engine/strat_engine/filesystem.rs]: Is there more that this method should be doing, i.e., is there a TODO that belongs here?

**Comment by agrover** [comment]
The filesystem opens the block device and takes a reference to it in the kernel until it's unmounted. This prevents the block device from being destroyed until the filesystem is unmounted.

File: src/engine/strat_engine/pool.rs (hunk 1 above)
Comment (agrover) [inline src/engine/strat_engine/pool.rs]: if you have filesystems A-C and B is still mounted, then it'll cause an error and teardown will return, leaving things in a partially torn-down state, yes.

File: src/engine/strat_engine/filesystem.rs (hunk 2 above)
Comment (agrover) [inline src/engine/strat_engine/filesystem.rs]: I dunno, like what?

Comment (mulkieran) [inline src/engine/strat_engine/filesystem.rs]: I don't know much about this area. You are saying that you are pretty sure there is nothing?

File: src/engine/strat_engine/pool.rs (hunk 1 above)
Comment (mulkieran) [inline src/engine/strat_engine/pool.rs]: Is there anyway we can detect if any are mounted before tearing down the filesystems?

Comment (agrover) [inline src/engine/strat_engine/pool.rs]: we can look at /proc/mounts ...

But, what's the use case here? Can we say teardown is only going to be called when the system is shutting down? (Teardown is not exposed via API.)

File: src/engine/strat_engine/filesystem.rs (hunk 2 above)
Comment (agrover) [inline src/engine/strat_engine/filesystem.rs]: I can't think of anything, or else I would've added it or added a TODO 😃

File: src/engine/strat_engine/pool.rs (hunk 1 above)
Comment (mulkieran) [inline src/engine/strat_engine/pool.rs]: We can not say that.

File: src/engine/strat_engine/filesystem.rs (hunk 2 above)
Comment (mulkieran) [inline src/engine/strat_engine/filesystem.rs]: Ok!

File: src/engine/strat_engine/pool.rs (hunk 1 above)
Comment (mulkieran) [inline src/engine/strat_engine/pool.rs]: Ok, please get rid of the note and put something like this in the function header comment:

```Precondition: All filesystems belonging to this pool must be unmounted.```

**Comment by agrover** [comment]
Made changes based on feedback.

File: src/engine/strat_engine/pool.rs [hunk 3]
```diff
@@ -376,13 +376,15 @@ impl StratPool {
...
+    /// Precondition: All filesystems belonging to this pool must be
+    /// unmounted.
     pub fn teardown(self) -> EngineResult<()> {
-        // TODO: any necessary clean up of filesystems
-        if !self.filesystems.is_empty() {
-            return Err(EngineError::Engine(ErrorEnum::Busy,
-                                           format!("May be unsynced files on device.")));
-        }
         let dm = try!(DM::new());
+
```
Comment (mulkieran) [inline src/engine/strat_engine/pool.rs]: You'll need to add the same action to ```destroy()```, I think, if so, same precondition.

Comment (agrover) [inline src/engine/strat_engine/pool.rs]: ok will do

**Comment by agrover** [comment]
updated.

**Pull request merged**